integration-tests/tests/src/tests/sync/user.ts

```diff
@@ -0,0 +1,620 @@
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+import { expect } from "chai";
+import { importAppBefore } from "../../hooks";
+import {
+  randomNonVerifiableEmail,
+  randomPendingVerificationEmail,
+  randomVerifiableEmail,
+  uuid,
+} from "../../utils/generators";
+import { KJUR } from "jsrsasign";
+import { sleep } from "../../utils/sleep";
+
+function assertIsUser(user: Realm.User) {
+  expect(user).to.not.be.undefined;
+  expect(user).to.not.be.null;
+  expect(typeof user).equals("object");
+  expect(typeof user.accessToken).equals("string");
+  expect(typeof user.refreshToken).equals("string");
+  expect(typeof user.id).equals("string");
+  expect(typeof user.identities).equals("object");
+  expect(typeof user.customData).equals("object");
+  expect(user).instanceOf(Realm.User);
+}
+
+function assertIsSameUser(value: Realm.User, user: Realm.User | null) {
+  assertIsUser(value);
+  expect(value.accessToken).equals(user?.accessToken);
+  expect(value.id).equals(user?.id);
+}
+
+function assertIsError(error: any, message: string) {
+  expect(error).instanceOf(Error, "The API should return an Error");
+  if (message) {
+    expect(error.message).equals(message);
+  }
+}
+
+function assertIsAuthError(error: any, code: number, title: string) {
+  expect(error).instanceOf(Realm.App.Sync.AuthError, "The API should return an AuthError");
+  if (code) {
+    expect(error.code).equals(code);
+  }
+  if (title) {
+    expect(error.title).equals(title);
+  }
+}
+
+async function registerAndLogInEmailUser(app: Realm.App) {
+  const validEmail = randomVerifiableEmail();
+  const validPassword = "test1234567890";
+  await app.emailPasswordAuth.registerUser({ email: validEmail, password: validPassword });
+  const user = await app.logIn(Realm.Credentials.emailPassword(validEmail, validPassword));
+  assertIsUser(user);
+  assertIsSameUser(user, app.currentUser);
+  return user;
+}
+
+async function removeExistingUsers(app: Realm.App) {
+  const users = app.allUsers;
+  Object.keys(app.allUsers).forEach(async (id) => {
+    await app.removeUser(users[id]);
+  });
+}
+
+describe.skipIf(environment.missingServer, "User", () => {
+  describe("email password", () => {
+    importAppBefore("with-email-password");
+    it("login without username throws", async function (this: Mocha.Context & AppContext & RealmContext) {
```

Comment:
While this is correct, both [`AppContext` and `RealmContext` unions with `Mocha.Context` underneath](https://github.com/realm/realm-js/blob/bindgen/integration-tests/tests/src/typings.d.ts#L68-L73), so there's no need to include that here.

```suggestion
    it("login without username throws", async function (this: AppContext & RealmContext) {
```